autopush_rs/src/util/megaphone.rs

```diff
@@ -0,0 +1,253 @@
+use std::collections::HashMap;
+use errors::Result;
+
+// A Service entry Key in a ServiceRegistry
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+struct ServiceKey(u32);
+
+// A list of services that a client is interested in and the last change seen
+#[derive(Debug)]
+pub struct ServiceList {
+    service_list: Vec<ServiceKey>,
+    change_count: u32,
+}
+
+#[derive(Debug)]
+struct ServiceRegistry {
+    lookup: HashMap<String, u32>,
+    table: Vec<String>,
+}
+
+// Return result of the first delta call for a client given a full list of service id's and versions
+#[derive(Debug)]
+pub struct ServiceClientInit {
+    pub service_list: ServiceList,
+    pub delta: Vec<Service>,
+}
+
+impl ServiceRegistry {
+    fn new() -> ServiceRegistry {
+        ServiceRegistry {
+            lookup: HashMap::new(),
+            table: Vec::new(),
+        }
+    }
+
+    // Add's a new service to the lookup table, returns the existing key if the service already
+    // exists
+    fn add_service(&mut self, service_id: String) -> ServiceKey {
+        if let Some(v) = self.lookup.get(&service_id) {
+            return ServiceKey(v.clone());
+        }
+        let i = self.table.len();
+        self.table.push(service_id.clone());
+        self.lookup.insert(service_id, i as u32);
+        ServiceKey(i as u32)
+    }
+
+    fn lookup_id(&self, key: ServiceKey) -> Option<String> {
+        if key.0 >= (self.table.len() as u32) {
+            None
+        } else {
+            Some(self.table[key.0 as usize].clone())
+        }
+    }
+
+    fn lookup_key(&self, service_id: String) -> Option<ServiceKey> {
+        self.lookup.get(&service_id).and_then(|v| Some(ServiceKey(v.clone())))
```

Comment:
`.and_then(|x| Some(...))` should be the same as `map(|x| ...)`

Reply:
although like above this could be `self.lookup.get(&service_id).cloned().map(ServiceKey)` (depending on preferences of course)

Reply:
I like that last one!